transport/transport.go

```diff
@@ -132,10 +131,16 @@ func (r *recvBufferReader) Read(p []byte) (n int, err error) {
 	if r.err != nil {
 		return 0, r.err
 	}
-	defer func() { r.err = err }()
-	if r.last != nil && r.last.Len() > 0 {
+	n, r.err = r.read(p)
+	return n, r.err
+}
+
+func (r *recvBufferReader) read(p []byte) (n int, err error) {
+	if r.last != nil && len(r.last) > 0 {
```

Comment:
nit: Since last is a slice now, we can get away with checking only for len(r.last) > 0 
https://play.golang.org/p/8_3pzaZZWI